[RISCV] Macro-fusion support for veyron-v1 CPU.

mgudim opened the pull request:
Support was added for the following fusions:
  auipc-addi, slli-srli, ld-add
Some parts of the code became repetative, so small refactoring of existing lui-addi fusion was done.

mgudim committed:
[RISCV] Macro-fusion support for veyron-v1 CPU.

Support was added for the following fusions:
  auipc-addi, slli-srli, ld-add
Some parts of the code became repetative, so small refactoring of
existing lui-addi fusion was done.

wangpc-pp commented:
As far as I know, there may be up to 30+ instruction pairs that can be fused for some macroarchitecture (for example, , [Xiangshan - Decode Unit, Chinese](https://xiangshan-doc.readthedocs.io/zh_CN/latest/frontend/decode/)). If we define subtarget features for all of them, that would be a disaster I think. So, I'm working on a TableGen backend that generates predicators for macro fusion (with some refactor of `ScheduleModel`) and I will raise a RFC and PR later. This is still WIP and  discussions are welcome!

Anyway, this PR looks great to me, and I'm just sharing some information here.

(For current implementation, I don't know if we can refer to PPC's, in which there is a script-generated header.)

mgudim commented:
> As far as I know, there may be up to 30+ instruction pairs that can be fused for some macroarchitecture (for example, , [Xiangshan - Decode Unit, Chinese](https://xiangshan-doc.readthedocs.io/zh_CN/latest/frontend/decode/)). If we define subtarget features for all of them, that would be a disaster I think. So, I'm working on a TableGen backend that generates predicators for macro fusion (with some refactors of `ScheduleModel`) and I will raise a RFC and PR later. This is still WIP and discussions are welcome!
> 
> Anyway, this PR looks great to me, and I'm just sharing some information here.
> 
> (For current implementation, I don't know if we can refer to PPC's, in which there is a script-generated header.)

Thanks for taking a look. I agree that we shouldn't have too many features defined. I'm happy to rework this in the future if we adopt a new approach.

topperc commented on llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:
```diff
@@ -18,6 +18,90 @@
+  return checkRegisters(FirstMI->getOperand(0).getReg(), SecondMI);
+}
+
+// Fuse SLLI by 32 feeding into SRLI by 32 or less or
```
Can we split this into two separate sentences? The `32 or less or` is kind of hard to parse.

mgudim replied:
I've replaced this with pseudocode.

topperc commented on llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:
```diff
@@ -27,29 +111,18 @@ static bool isLUIADDI(const MachineInstr *FirstMI,
-
-  // Destination of LUI should be the ADDI(W) source register.
-  if (SecondMI.getOperand(1).getReg() != FirstDest)
+  // The first operand of ADDI might be a frame index.
```
This check was removed previously because it wasn't needed. Was this a bad merge or is it needed now?

mgudim replied:
In this code:

```
$rd0 =  lui $imm0
$rd1 = addi $rs0, $imm1
```

In order for fusion to happen it must be that `$rd0 == rd1` and `$rd0 == rs0`. Same for other fusions. Both checks are done in `checkRegisters` now. I think both checks were needed before two, according to this comment:

```
// Fuse LUI followed by ADDI or ADDIW.
// rd = imm[31:0] which decomposes to
// lui rd, imm[31:12]
// addi(w) rd, rd, imm[11:0]

```

topperc replied:
I was specifically refering to the FrameIndex comment.

mgudim replied:
I think we do need the check `SecondMI.getOperand(1).isReg()`  because later we do `SecondMI.getOperand(1).getReg() != FirstDest` - this may crash, right?

I've put that check as a first line of `checkRegisters`.

topperc replied:
As was explained in #68701, if the ADDI operand isn't a register then there can be no data dependency between the LUI and the ADDi so it wouldn't be considered for macrofusion.

mgudim replied:
oh, sorry I didn't realize that by the time we call `shouldScheduleAdjacent` we know that there is a data dependency. Ok, I looked at  `MacroFusion.cpp : 177` - we only consider pairs of vertices connected by an edge of a schedule dag. But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

wangpc-pp replied:
> But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

Yes, I have to say that it's possible to have other order dependencies like `Artificial`. For example, `BaseMemOpClusterMutation::clusterNeighboringMemOps`. Though we may not add artificial edges (not for fusion) between ALU instructions. I haven't run into any problem after remove the `isReg()` yet, but I think we may add it back since there are some potential risks of assertion theoretically.

mgudim committed:
addressed review comments.

preames commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -242,4 +242,5 @@ def VENTANA_VEYRON_V1 : RISCVProcessorModel<"veyron-v1",
                                              FeatureStdExtZicboz,
-                                             FeatureVendorXVentanaCondOps]>;
+                                             FeatureVendorXVentanaCondOps,
+                                             TuneVeyronFusions]>;
```
It looks like you're adding this to the feature list, not the tune list.  That's probably not what you meant.

mgudim replied:
done in https://github.com/llvm/llvm-project/pull/70414

preames commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -954,6 +964,14 @@ def TuneSiFive7 : SubtargetFeature<"sifive7", "RISCVProcFamily", "SiFive7",
                                    [TuneNoDefaultUnroll,
                                     TuneShortForwardBranchOpt]>;
 
+def TuneVeyronFusions : SubtargetFeature<"ventana-veyron", "RISCVProcFamily", "VentanaVeyron",
```
Please separate this out in a separate PR which only adds the proc family and the existing LUI/ADDI fusion.  Then this change can extend the list with the new fusions.

mgudim replied:
done: https://github.com/llvm/llvm-project/pull/70414

preames commented on llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:
```diff
@@ -18,6 +18,97 @@
+    return false;
+
+  unsigned SRLIImm = SecondMI.getOperand(2).getImm();
+  bool IsShiftBy48 = SRLIImm == 48;
```
This seems a very oddly restricted fusion.  Are you sure that e.g. a shift pair by 37 that clears the top bits isn't fused?

topperc replied:
The two shifts by 48 case is the i16 zext pattern for rv64 without Zbb. So it makes some sense.

preames commented on llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:
```diff
@@ -18,6 +18,97 @@
+  if (FirstMI->getOpcode() != RISCV::SLLI)
+    return false;
+
+  unsigned SLLIImm = FirstMI->getOperand(2).getImm();
```
Are intending to allow two different shift amounts less than 32?  That seems unlikely.

mgudim replied:
Yes, you're right, thanks. Fixed that.

preames reviewed (commented):
Is there an optimization guide or other public reference which describes these fusions?

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -937,6 +937,16 @@ def TuneLUIADDIFusion
+def TuneAUIPCADDIFusion
+    : SubtargetFeature<"auipc-addi-fusion", "HasAUIPCADDIFusion",
+                       "true", "Enable AUIPC+ADDI macrofusion">;
+def TuneSLLISRLIFusion
```
This is pretty generic name given the specific cases that are recognized for fusion.

mgudim replied:
how about `ShiftedZExtFusion`?

topperc replied:
ok

mgudim committed:
Corrected slli-srli fusion - the immediate shift ammounts have to be the
same.

asb commented:
Could we add tests for this please? I know we have macro-fusion-lui-addi.ll already though I'm not sure .ll tests are the best for this. PowerPC's macro-fusion.mir test seemed like a nice way to handle it, to my eye at least.

mgudim committed:
Added test.

mgudim commented:
> Could we add tests for this please? I know we have macro-fusion-lui-addi.ll already though I'm not sure .ll tests are the best for this. PowerPC's macro-fusion.mir test seemed like a nice way to handle it, to my eye at least.

Yes, good idea. Done.

mgudim commented:
@asb @preames @topperc Ping

mgudim committed:
Renamed SLLISRLIFusion into ShiftedZExtFusion.
Added pseudocode to explain fusions.
Replace the "spacer" `ADDI` in the test with `XORI`- it was
confusing in case of LDADD fusion.

wangpc-pp commented on llvm/test/CodeGen/RISCV/macro-fusions-veyron-v1.mir:
```diff
@@ -0,0 +1,159 @@
+# REQUIRES: asserts
```
Do we need end-to-end llc tests like `llvm/test/CodeGen/RISCV/macro-fusion-lui-addi.ll`?

wangpc-pp reviewed (commented):
I'd like to see this landed so that there are more use cases for TableGen way. :-)

mgudim commented:
@topperc @asb @preames I think all the comments are addressed, sorry for taking so long. Is it OK to merge?

topperc reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:
```diff
@@ -18,6 +18,101 @@
+// add rd, rs1, rs2
+// ld rd, 0(rd)
+static bool isLDADD(const MachineInstr *FirstMI, const MachineInstr &SecondMI) {
+  if (SecondMI.getOpcode() != RISCV::LD)
```
@mgudim Is this fusion reall restricted to just 64-bit load?

mgudim replied:
yes.
